// http-react/dist/internal/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UNITS_MILISECONDS_EQUIVALENTS = exports.METHODS = exports.ALLOWED_CONTEXT_KEYS = exports.DEFAULT_RESOLVER = exports.DEFAULT_GRAPHQL_PATH = exports.REVALIDATE_ON_MOUNT = exports.RETRY_ON_RECONNECT = exports.REVALIDATE_ON_FOCUS = exports.ATTEMPT_INTERVAL = exports.ATTEMPTS = exports.ONLINE = exports.ON_ONLINE = exports.ON_OFFLINE = exports.PARAMS = exports.QUERY = exports.DEFAULTS = void 0;
var DEFAULTS = {};
exports.DEFAULTS = DEFAULTS;
var QUERY = {};
exports.QUERY = QUERY;
var PARAMS = {};
exports.PARAMS = PARAMS;
var ON_OFFLINE = function () { };
exports.ON_OFFLINE = ON_OFFLINE;
var ON_ONLINE = function () { };
exports.ON_ONLINE = ON_ONLINE;
var ONLINE = true;
exports.ONLINE = ONLINE;
var ATTEMPTS = 0;
exports.ATTEMPTS = ATTEMPTS;
var ATTEMPT_INTERVAL = 2;
exports.ATTEMPT_INTERVAL = ATTEMPT_INTERVAL;
var REVALIDATE_ON_FOCUS = false;
exports.REVALIDATE_ON_FOCUS = REVALIDATE_ON_FOCUS;
var RETRY_ON_RECONNECT = true;
exports.RETRY_ON_RECONNECT = RETRY_ON_RECONNECT;
var REVALIDATE_ON_MOUNT = true;
exports.REVALIDATE_ON_MOUNT = REVALIDATE_ON_MOUNT;
var DEFAULT_GRAPHQL_PATH = '/graphql';
exports.DEFAULT_GRAPHQL_PATH = DEFAULT_GRAPHQL_PATH;
var DEFAULT_RESOLVER = function (e) { return e.json(); };
exports.DEFAULT_RESOLVER = DEFAULT_RESOLVER;
var ALLOWED_CONTEXT_KEYS = [
    'headers',
    'baseUrl',
    'body',
    'defaults',
    'resolver',
    'auto',
    'memory',
    'refresh',
    'attempts',
    'attemptInterval',
    'revalidateOnFocus',
    'query',
    'params',
    'onOnline',
    'onOffline',
    'online',
    'retryOnReconnect',
    'cache'
];
exports.ALLOWED_CONTEXT_KEYS = ALLOWED_CONTEXT_KEYS;
var METHODS = {
    GET: 'GET',
    DELETE: 'DELETE',
    HEAD: 'HEAD',
    OPTIONS: 'OPTIONS',
    POST: 'POST',
    PUT: 'PUT',
    PATCH: 'PATCH',
    PURGE: 'PURGE',
    LINK: 'LINK',
    UNLINK: 'UNLINK'
};
exports.METHODS = METHODS;
var UNITS_MILISECONDS_EQUIVALENTS = {
    ms: 1,
    sec: 1000,
    min: 60000,
    h: 3600000,
    d: 86400000,
    we: 604800000,
    mo: 2629800000,
    y: 31536000000
};
exports.UNITS_MILISECONDS_EQUIVALENTS = UNITS_MILISECONDS_EQUIVALENTS;
