// http-react/dist/hooks/others.js
'use client';
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useImperative = exports.useGql = exports.useDebounceFetch = exports.useRequestEnd = exports.useRequestStart = exports.useFetchResponseTime = exports.useFetchText = exports.useFetchBlob = exports.useUNLINK = exports.useLINK = exports.usePURGE = exports.usePATCH = exports.usePUT = exports.usePOST = exports.useOPTIONS = exports.useHEAD = exports.useDELETE = exports.useGET = exports.useResolve = exports.useFetchId = exports.useSuccess = exports.useHasData = exports.useExpiration = exports.useRevalidating = exports.useReFetch = exports.useLoadingFirst = exports.useOnline = exports.useFetchMutate = exports.useFetchError = exports.useFetchLoading = exports.useFetchCode = exports.useFetchData = exports.useFetchSuspense = exports.useFetchConfig = void 0;
var React = __importStar(require("react"));
var react_1 = require("react");
var internal_1 = require("../internal");
var use_fetch_1 = require("./use-fetch");
var utils_1 = require("../utils");
var shared_1 = require("../utils/shared");
var constants_1 = require("../internal/constants");
/**
 * Get the current fetcher config
 */
function useFetchConfig(id) {
    var ftxcf = (0, internal_1.useHRFContext)();
    var config = (0, use_fetch_1.useFetch)({ id: id }).config;
    var thisConfig = (0, shared_1.isDefined)(id) ? config : ftxcf;
    return thisConfig;
}
exports.useFetchConfig = useFetchConfig;
function useFetchSuspense(init, options) {
    var o = typeof init === 'string'
        ? __assign({ url: init }, options) : options;
    return (0, use_fetch_1.useFetch)(__assign(__assign({}, o), { suspense: true }));
}
exports.useFetchSuspense = useFetchSuspense;
/**
 * Get the data state of a request using its id
 */
function useFetchData(id, onResolve) {
    var _a = (0, internal_1.useHRFContext)().cacheProvider, cacheProvider = _a === void 0 ? internal_1.defaultCache : _a;
    var defaultsKey = (0, shared_1.serialize)({
        idString: (0, shared_1.serialize)(id)
    });
    var def = cacheProvider.get(defaultsKey);
    var data = (0, use_fetch_1.useFetch)({
        default: def,
        id: id,
        onResolve: onResolve
    }).data;
    return data;
}
exports.useFetchData = useFetchData;
function useFetchCode(id) {
    var code = (0, use_fetch_1.useFetch)({
        id: id
    }).code;
    return code;
}
exports.useFetchCode = useFetchCode;
/**
 * Get the loading state of a request using its id
 */
function useFetchLoading(id) {
    var _a = (0, use_fetch_1.useFetch)({
        id: id
    }), loading = _a.loading, all = __rest(_a, ["loading"]);
    return (loading &&
        (0, internal_1.isPending)((0, shared_1.serialize)({
            idString: (0, shared_1.serialize)(id)
        })));
}
exports.useFetchLoading = useFetchLoading;
/**
 * Get the error state of a request using its id
 */
function useFetchError(id, onError) {
    var resolvedKey = (0, shared_1.serialize)({
        idString: (0, shared_1.serialize)(id)
    });
    (0, react_1.useEffect)(function () {
        function listenToErrorEvent(e) {
            if (e === null || e === void 0 ? void 0 : e.error) {
                if ((0, shared_1.isFunction)(onError)) {
                    ;
                    onError(e === null || e === void 0 ? void 0 : e.error);
                }
            }
        }
        internal_1.requestsProvider.addListener(resolvedKey, listenToErrorEvent);
        return function () {
            internal_1.requestsProvider.removeListener(resolvedKey, listenToErrorEvent);
        };
    }, [resolvedKey]);
    return ((0, use_fetch_1.useFetch)({
        id: id
    }).error || internal_1.hasErrors[resolvedKey]);
}
exports.useFetchError = useFetchError;
/**
 * Get the mutate the request data using its id
 */
function useFetchMutate(
/**
 * The id of the `useFetch` call
 */
id, 
/**
 * The function to run after mutating
 */
onMutate) {
    var mutate = (0, use_fetch_1.useFetch)({
        id: id,
        onMutate: onMutate
    }).mutate;
    return mutate;
}
exports.useFetchMutate = useFetchMutate;
function useOnline(id) {
    return (0, use_fetch_1.useFetch)({ id: id }).online;
}
exports.useOnline = useOnline;
function useLoadingFirst(id) {
    return (0, use_fetch_1.useFetch)({ id: id }).loadingFirst;
}
exports.useLoadingFirst = useLoadingFirst;
function useReFetch(id) {
    return (0, use_fetch_1.useFetch)({ id: id }).reFetch;
}
exports.useReFetch = useReFetch;
function useRevalidating(id) {
    return (0, use_fetch_1.useFetch)({ id: id }).revalidating;
}
exports.useRevalidating = useRevalidating;
function useExpiration(id) {
    return (0, use_fetch_1.useFetch)({ id: id }).expiration;
}
exports.useExpiration = useExpiration;
function useHasData(id) {
    return (0, use_fetch_1.useFetch)({ id: id }).hasData;
}
exports.useHasData = useHasData;
function useSuccess(id) {
    return (0, use_fetch_1.useFetch)({ id: id }).success;
}
exports.useSuccess = useSuccess;
/**
 * Get everything from a `useFetch` call using its id
 */
function useFetchId(id) {
    var defaultsKey = (0, shared_1.serialize)({
        idString: (0, shared_1.serialize)(id)
    });
    var def = internal_1.fetcherDefaults[defaultsKey];
    return (0, use_fetch_1.useFetch)({
        id: id,
        default: def
    });
}
exports.useFetchId = useFetchId;
/**
 * Create an effect for when the request completes
 */
function useResolve(id, onResolve) {
    var _a;
    var defaultsKey = (0, shared_1.serialize)({
        idString: (0, shared_1.serialize)(id)
    });
    (0, react_1.useEffect)(function () {
        function resolve(v) {
            return __awaiter(this, void 0, void 0, function () {
                var isResolved, data;
                return __generator(this, function (_a) {
                    isResolved = v.isResolved, data = v.data;
                    if (isResolved) {
                        if ((0, shared_1.isFunction)(onResolve)) {
                            onResolve(data, internal_1.lastResponses[defaultsKey]);
                        }
                    }
                    return [2 /*return*/];
                });
            });
        }
        internal_1.requestsProvider.addListener(defaultsKey, resolve);
        return function () {
            internal_1.requestsProvider.removeListener(defaultsKey, resolve);
        };
    }, [defaultsKey, onResolve]);
    return (_a = (0, use_fetch_1.useFetch)({ id: id })) === null || _a === void 0 ? void 0 : _a.data;
}
exports.useResolve = useResolve;
/**
 * User a `GET` request
 */
function useGET(init, options) {
    return (0, use_fetch_1.useFetch)(init, __assign(__assign({}, options), { method: 'GET' }));
}
exports.useGET = useGET;
/**
 * Use a `DELETE` request
 */
function useDELETE(init, options) {
    return (0, use_fetch_1.useFetch)(init, __assign(__assign({}, options), { method: 'DELETE' }));
}
exports.useDELETE = useDELETE;
/**
 * Use a `HEAD` request
 */
function useHEAD(init, options) {
    return (0, use_fetch_1.useFetch)(init, __assign(__assign({}, options), { method: 'HEAD' }));
}
exports.useHEAD = useHEAD;
/**
 * Use an `OPTIONS` request
 */
function useOPTIONS(init, options) {
    return (0, use_fetch_1.useFetch)(init, __assign(__assign({}, options), { method: 'OPTIONS' }));
}
exports.useOPTIONS = useOPTIONS;
/**
 * Use a `POST` request
 */
function usePOST(init, options) {
    return (0, use_fetch_1.useFetch)(init, __assign(__assign({}, options), { method: 'POST' }));
}
exports.usePOST = usePOST;
/**
 * Use a `PUT` request
 */
function usePUT(init, options) {
    return (0, use_fetch_1.useFetch)(init, __assign(__assign({}, options), { method: 'PUT' }));
}
exports.usePUT = usePUT;
/**
 * Use a `PATCH` request
 */
function usePATCH(init, options) {
    return (0, use_fetch_1.useFetch)(init, __assign(__assign({}, options), { method: 'PATCH' }));
}
exports.usePATCH = usePATCH;
/**
 * Use a `PURGE` request
 */
function usePURGE(init, options) {
    return (0, use_fetch_1.useFetch)(init, __assign(__assign({}, options), { method: 'PURGE' }));
}
exports.usePURGE = usePURGE;
/**
 * Use a `LINK` request
 */
function useLINK(init, options) {
    return (0, use_fetch_1.useFetch)(init, __assign(__assign({}, options), { method: 'LINK' }));
}
exports.useLINK = useLINK;
/**
 * Use an `UNLINK` request
 */
function useUNLINK(init, options) {
    return (0, use_fetch_1.useFetch)(init, __assign(__assign({}, options), { method: 'UNLINK' }));
}
exports.useUNLINK = useUNLINK;
/**
 * Get a blob of the response. You can pass an `objectURL` property that will convet that blob into a string using `URL.createObjectURL`
 */
function useFetchBlob(init, options) {
    return (0, use_fetch_1.useFetch)(init, __assign(__assign({}, options), { resolver: function (res) {
            return __awaiter(this, void 0, void 0, function () {
                var blob;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, res.blob()];
                        case 1:
                            blob = _a.sent();
                            if (typeof URL !== 'undefined') {
                                if (init.objectURL) {
                                    return [2 /*return*/, URL.createObjectURL(blob)];
                                }
                                else {
                                    if (options === null || options === void 0 ? void 0 : options.objectURL) {
                                        return [2 /*return*/, URL.createObjectURL(blob)];
                                    }
                                }
                            }
                            return [2 /*return*/, blob];
                    }
                });
            });
        } }));
}
exports.useFetchBlob = useFetchBlob;
/**
 * Get a text of the response
 */
function useFetchText(init, options) {
    return (0, use_fetch_1.useFetch)(init, __assign(__assign({}, options), { resolver: function (res) {
            return __awaiter(this, void 0, void 0, function () {
                var text;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, res.text()];
                        case 1:
                            text = _a.sent();
                            return [2 /*return*/, text];
                    }
                });
            });
        } }));
}
exports.useFetchText = useFetchText;
function useFetchResponseTime(id) {
    return (0, use_fetch_1.useFetch)({
        id: id
    }).responseTime;
}
exports.useFetchResponseTime = useFetchResponseTime;
/**
 * Get the `Date` the request started
 */
function useRequestStart(id) {
    return (0, use_fetch_1.useFetch)({
        id: id
    }).requestStart;
}
exports.useRequestStart = useRequestStart;
/**
 * Get the `Date` the request finished
 */
function useRequestEnd(id) {
    return (0, use_fetch_1.useFetch)({
        id: id
    }).requestEnd;
}
exports.useRequestEnd = useRequestEnd;
/**
 * Debounce a fetch by the time given
 */
function useDebounceFetch(init, options) {
    var _a, _b;
    // @ts-ignore - auto can be present in the first arg
    var canDebounce = (_b = (_a = init === null || init === void 0 ? void 0 : init.auto) !== null && _a !== void 0 ? _a : options === null || options === void 0 ? void 0 : options.auto) !== null && _b !== void 0 ? _b : true;
    var res = (0, use_fetch_1.useFetch)(init, __assign(__assign({}, options), { auto: false }));
    // @ts-ignore - debounce can be present in the first arg
    var debounce = (0, utils_1.getMiliseconds)((init === null || init === void 0 ? void 0 : init.debounce) || (options === null || options === void 0 ? void 0 : options.debounce) || '0 ms');
    (0, react_1.useEffect)(function () {
        var debounceTimeout = null;
        if (canDebounce) {
            debounceTimeout = setTimeout(res.reFetch, debounce);
        }
        return function () {
            clearTimeout(debounceTimeout);
        };
    }, [(0, shared_1.serialize)({ init: init, options: options }), canDebounce]);
    return res;
}
exports.useDebounceFetch = useDebounceFetch;
/**
 * Make a graphQL request
 */
function useGql(arg1, cfg) {
    if (cfg === void 0) { cfg = {}; }
    var isUsingExternalQuery = typeof arg1.value === 'string';
    var query;
    if (isUsingExternalQuery) {
        query = arg1.value;
    }
    else {
        query = arg1[0][0];
    }
    var _a = cfg.variables, variables = _a === void 0 ? {} : _a, _b = cfg.graphqlPath, graphqlPath = _b === void 0 ? constants_1.DEFAULT_GRAPHQL_PATH : _b, otherArgs = __rest(cfg, ["variables", "graphqlPath"]);
    var config = otherArgs;
    var JSONBody = (0, shared_1.serialize)({
        query: query,
        variables: variables
    });
    var usingProvider = (0, shared_1.isDefined)(cfg['__fromProvider']);
    var g = (0, use_fetch_1.useFetch)(__assign(__assign(__assign(__assign(__assign(__assign(__assign({}, config), { id: arg1 }), { variables: cfg.variables || {} }), otherArgs), { default: usingProvider
            ? otherArgs.default
            : {
                data: cfg === null || cfg === void 0 ? void 0 : cfg.default,
                errors: undefined,
                variables: cfg.variables || {}
            } }), { __gql: true }), { params: {
            // Caching versions of different variables values
            variables: variables
        }, formatBody: function () { return JSONBody; }, body: JSONBody, method: 'POST', url: graphqlPath }));
    return g;
}
exports.useGql = useGql;
/**
 * Use an imperative version of the fetcher (similarly to Axios, it returns an object with `get`, `post`, etc)
 */
function useImperative() {
    var ctx = useFetchConfig();
    var imperativeFetch = React.useMemo(function () { return (0, utils_1.createImperativeFetch)(ctx); }, [(0, shared_1.serialize)(ctx)]);
    return imperativeFetch;
}
exports.useImperative = useImperative;
